File: frontend/src/lib/permissions.ts
import type { RolUsuario } from "@/types";

export type Capability =
  | "proyectos.view"
  | "proyectos.edit"
  | "proyectos.change_state"
  | "diputados.view"
  | "comisiones.view"
  | "documentos.download"
  | "bitacora.view";

const ROLE_CAPABILITIES: Record<RolUsuario, Capability[]> = {
  ciudadano: [
    "proyectos.view",
    "diputados.view",
    "comisiones.view",
    "documentos.download",
  ],
  diputado: [
    "proyectos.view",
    "diputados.view",
    "comisiones.view",
    "documentos.download",
  ],
  admin: [
    "proyectos.view",
    "proyectos.edit",
    "proyectos.change_state",
    "diputados.view",
    "comisiones.view",
    "documentos.download",
    "bitacora.view",
  ],
};

export function getRoleCapabilities(role: RolUsuario | null | undefined): Capability[] {
  if (!role) return ROLE_CAPABILITIES.ciudadano;
  return ROLE_CAPABILITIES[role] ?? ROLE_CAPABILITIES.ciudadano;
}

export function hasCapability(
  role: RolUsuario | null | undefined,
  capability: Capability
): boolean {
  return getRoleCapabilities(role).includes(capability);
}

export function canEditProyecto(role: RolUsuario | null | undefined): boolean {
  return hasCapability(role, "proyectos.edit");
}

export function canChangeProyectoEstado(role: RolUsuario | null | undefined): boolean {
  return hasCapability(role, "proyectos.change_state");
}

export function canViewBitacora(role: RolUsuario | null | undefined): boolean {
  return hasCapability(role, "bitacora.view");
}

File: gobi-backend/app/core/permissions.py
from fastapi import HTTPException, Security
from typing import Callable

from app.core.auth import get_current_user

Capability = str

ROLE_CAPABILITIES = {
    "ciudadano": {
        "proyectos.view",
        "diputados.view",
        "comisiones.view",
        "documentos.download",
    },
    "diputado": {
        "proyectos.view",
        "diputados.view",
        "comisiones.view",
        "documentos.download",
    },
    "admin": {
        "proyectos.view",
        "proyectos.edit",
        "proyectos.change_state",
        "diputados.view",
        "comisiones.view",
        "documentos.download",
        "bitacora.view",
    },
}


def has_capability(user: dict, capability: Capability) -> bool:
    rol = user.get("rol", "ciudadano")
    capabilities = ROLE_CAPABILITIES.get(rol, ROLE_CAPABILITIES["ciudadano"])
    return capability in capabilities


def require_capability(capability: Capability) -> Callable:
    async def dependency(user=Security(get_current_user)):
        if not has_capability(user, capability):
            raise HTTPException(
                status_code=403,
                detail=f"No tienes permiso para: {capability}"
            )
        return user

    return dependency
